File: Algorithms/Python3.x/678-Valid_Parenthesis_String.py
# Greedy
# Runtime: 32 ms, faster than 54.10% of Python3 online submissions for Valid Parenthesis String.
# Memory Usage: 13.7 MB, less than 84.78% of Python3 online submissions for Valid Parenthesis String.
class Solution:
    def checkValidString(self, s: str) -> bool:
        low, high = 0, 0
        for c in s:
            low  += 1 if c == "(" else -1
            high += 1 if c != ")" else -1
            if high < 0:
                break
            low = max(low, 0)
        return low == 0

File: Algorithms/C/678-Valid_Parenthesis_String.c
// greedy
// Runtime: 0 ms, faster than 100.00% of C online submissions for Valid Parenthesis String.
// Memory Usage: 5.1 MB, less than 70.69% of C online submissions for Valid Parenthesis String.
bool checkValidString(char * s){
    int low  = 0;
    int high = 0;
    for (char* t = s; *t != '\0'; ++t) {
        if (*t == '(')
            low++;
        else
            low--;
        if (*t != ')')
            high++;
        else
            high--;
        if (high < 0)
            break;
        low = low > 0 ? low : 0;
    }
    return low == 0;
}
